### PR TITLE
feat: add leg ranking and settlement

### PR DESCRIPTION
**Pull request by hellkonig**
## Outcome

Add deterministic racing-camel ranking and immutable leg settlement using the betting state introduced in PR 5a.

## Changes

- rank racing camels first-to-last while excluding crazy camels and respecting stack levels and finish zones
- settle printed leg-ticket payouts, second-place payouts, losing-ticket penalties, and pyramid-ticket income
- floor player balances at zero and reset only leg-scoped betting assets
- preserve dice, spectator tiles, turn fields, final bets, camel positions, and terminal status
- expose the scoring API and document the completed PR 5b milestone

## Non-goals

- final winner or loser settlement
- dice or spectator-tile reset
- leg-number or current-player advancement
- legal actions, CLI migration, agents, or RL environments

## Verification

- 107 tests passed
- Ruff lint passed
- Ruff format check passed
- mypy passed
- git diff check passed

### CODE REVIEW

File: src/camel_up/engine/scoring.py [hunk 1]
```diff
@@ -0,0 +1,119 @@
+"""Deterministic racing-camel ranking and leg settlement rules."""
+
+from __future__ import annotations
```
Comment (hellkonig) [inline src/camel_up/engine/scoring.py]: Is it okay to import from `__future__` which looks like a private thing?

File: src/camel_up/engine/scoring.py [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+    """
+    if not all(position.is_placed for position in board.camel_positions):
+        raise ValueError("initial setup must be completed before ranking camels")
+
+    return tuple(
+        sorted(
+            RACING_CAMEL_ORDER,
+            key=lambda camel: _race_progress(board, camel),
+            reverse=True,
+        )
```
Comment (hellkonig) [inline src/camel_up/engine/scoring.py]: Can you explain it? I am not familiar with the `lambda` function.

File: src/camel_up/engine/scoring.py [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+        players=players,
+        available_leg_betting_tickets=INITIAL_LEG_BETTING_TICKET_STACKS,
+    )
+
+
+def _race_progress(board: BoardState, camel: CamelId) -> tuple[int, int]:
+    """Return a sortable racing progress coordinate for one placed camel."""
+    position = position_of(board, camel)
+    if position.space is None or position.level is None:
+        raise ValueError("initial setup must be completed before ranking camels")
```
Comment (hellkonig) [inline src/camel_up/engine/scoring.py]: Maybe explicitly say which camel's position is `None`; it will be more informative.

File: src/camel_up/engine/scoring.py [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+        leg_betting_tickets=(),
+    )
+
+
+def _validate_leg_settlement(state: GameState) -> None:
+    """Reject states that have not reached a scoring boundary."""
+    if not all(position.is_placed for position in state.board.camel_positions):
+        raise ValueError("initial setup must be completed before settling a leg")
+    if not state.terminal and len(state.remaining_dice) != 1:
+        raise ValueError("a leg can only be settled after five dice have been used")
```
Comment (hellkonig) [inline src/camel_up/engine/scoring.py]: A leg can also be settled when the state is terminated, as the condition says. The `ValueError` is not completed; it may also explicitly say the current state will be helpful.

File: src/camel_up/engine/scoring.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+    """Return a sortable racing progress coordinate for one placed camel."""
+    position = position_of(board, camel)
+    if position.space is None or position.level is None:
+        raise ValueError("initial setup must be completed before ranking camels")
+    return position.space, position.level
+
+
+def _settle_player(
+    player: PlayerState,
+    *,
```
Comment (hellkonig) [inline src/camel_up/engine/scoring.py]: Why does this function have `*` here?

File: tests/engine/test_scoring.py [hunk 6]
```diff
@@ -0,0 +1,221 @@
...
+        CamelId.BLUE,
+        CamelId.YELLOW,
+        CamelId.PURPLE,
+    )
+
+
+def test_ranking_handles_forward_and_backward_finish_zone_stacks() -> None:
+    board = _board_with_stacks(
+        {
+            -1: (CamelId.BLACK, CamelId.PURPLE, CamelId.GREEN),
```
Comment (hellkonig) [inline tests/engine/test_scoring.py]: Why is there a `-1` space?

File: tests/engine/test_scoring.py [hunk 7]
```diff
@@ -0,0 +1,221 @@
...
+def test_leg_result_is_combined_before_money_is_floored_at_zero() -> None:
+    state = take_leg_betting_ticket(_active_state(), 0, CamelId.GREEN)
+    player = replace(state.players[0], money=0, pyramid_ticket_count=1)
+    state = replace(
+        state,
+        players=(player, *state.players[1:]),
+        remaining_dice=(DieId.YELLOW,),
+    )
+
+    settled = settle_leg(state)
```
Comment (hellkonig) [inline tests/engine/test_scoring.py]: This test tests that player 0 earns 1 from pyramid_ticket, but loses 1 from green ticket. Because the money is 0, so it is expected that the money after this leg is still 0. It is good, but it is also worth testing whether the money becomes negative; the player's money is still 0.

File: tests/engine/test_scoring.py [hunk 8]
```diff
@@ -0,0 +1,221 @@
...
+
+    # The losing ticket and pyramid ticket net to zero. Flooring the loss
+    # before adding the income would incorrectly leave the player with 1 EP.
+    assert settled.players[0].money == 0
+
+
+def test_leg_settlement_preserves_race_long_and_orchestration_state() -> None:
```
Comment (hellkonig) [inline tests/engine/test_scoring.py]: Why in this test is the ' settled ' state the same as the `state`?

File: README.md
Comment (hellkonig) [inline README.md]: Again, please review the `README.md` to see if it is necessary for this context by considering the audience of the `README.md`.

File: src/camel_up/engine/scoring.py (hunk 1 above)
Comment (hellkonig) [reply src/camel_up/engine/scoring.py]: Yes. __future__ is part of Python's standard library, not a private project module; the double underscores are simply its official module name. The annotations feature postpones type-annotation evaluation and is already used consistently across the engine modules. I kept it for package consistency and safe type-hint evaluation.

File: tests/engine/test_scoring.py (hunk 6 above)
Comment (hellkonig) [reply tests/engine/test_scoring.py]: Space -1 is the engine's canonical backward finish zone; BoardState.track_length is the forward finish zone. This is documented on CamelPosition and produced by backward finish crossings in movement.py. Commit 63b9e89 adds an inline test comment so the sentinel is clear in this scenario.

File: src/camel_up/engine/scoring.py (hunk 5 above)
Comment (hellkonig) [reply src/camel_up/engine/scoring.py]: The * makes first and second keyword-only arguments. Both have the same CamelId type, so requiring names at the call site prevents an accidental positional swap and makes payout code easier to review. I kept it as a small API-safety measure.

File: README.md
Comment (hellkonig) [reply README.md]: Agreed on keeping the README focused on its developer audience. Commit 63b9e89 removes the context-free scoring snippet and detailed reset list. It retains only a concise capability note so the README does not incorrectly say leg settlement is unavailable.

File: src/camel_up/engine/scoring.py (hunk 4 above)
Comment (hellkonig) [reply src/camel_up/engine/scoring.py]: Agreed. Commit 63b9e89 now reports that the state is non-terminal, includes the actual remaining-die count, and states both valid boundaries: one die remaining or a terminal game.

File: tests/engine/test_scoring.py (hunk 7 above)
Comment (hellkonig) [reply tests/engine/test_scoring.py]: Agreed. The existing net-zero case specifically proves that all gains and losses are aggregated before applying the floor, but it did not independently exercise a negative total. Commit 63b9e89 adds a separate negative-balance test that starts at zero with a losing ticket and confirms the result remains zero.

File: src/camel_up/engine/scoring.py (hunk 2 above)
Comment (hellkonig) [reply src/camel_up/engine/scoring.py]: The lambda is a small anonymous function used only as sorted's key. sorted calls it once per racing camel; it returns that camel's (space, stack level), which becomes the sort value. It is equivalent to defining a separate one-line local function. I kept the lambda because this short sort-key form is idiomatic Python and keeps the ranking logic next to the sort.

File: tests/engine/test_scoring.py (hunk 8 above)
Comment (hellkonig) [reply tests/engine/test_scoring.py]: The assertion was comparing settle_leg(settled) with settled, not the first settled state with the original input. Its purpose is to ensure repeated settlement cannot credit the same betting assets twice after those assets have been cleared. Commit 63b9e89 moves this into a dedicated, clearly named idempotence test with an explicit initial payout.

File: src/camel_up/engine/scoring.py (hunk 3 above)
Comment (hellkonig) [reply src/camel_up/engine/scoring.py]: Agreed that the defensive error should identify its subject. Commit 63b9e89 now reports the specific camel color. The public ranking path still rejects an incomplete setup first because BoardState intentionally permits either all camels placed or all unplaced, never a partial snapshot.

File: tests/engine/test_scoring.py (hunk 8 above)
Comment (hellkonig) [reply tests/engine/test_scoring.py]: No, what I mean is the unit test test_leg_settlement_preserves_race_long_and_orchestration_state()

Comment (hellkonig) [reply tests/engine/test_scoring.py]: Thanks for clarifying. In test_leg_settlement_preserves_race_long_and_orchestration_state, only the explicitly listed race-long and orchestration fields should equal their input values; the whole settled state should not equal the input. Commit 3a0ccd5 now proves that distinction first: the complete states differ, player 1 differs, and the shared leg-ticket supply differs. The following equality assertions then document the fields that leg settlement deliberately does not own.
